relay_messages.go

```diff
@@ -228,6 +238,40 @@ func (f lazyCallReq) Arg2Iterator() (arg2.KeyValIterator, error) {
 	return arg2.NewKeyValIterator(f.Payload[f.arg2StartOffset:f.arg2EndOffset])
 }
 
+func (f *lazyCallReq) Arg2Append(key, val []byte) error {
+	arg3 := bufPool.Get().(*bytes.Buffer)
+	arg3.Write(f.Payload[f.arg2EndOffset:])
+
+	// Pointer to arg2 payload
+	arg2Payload := f.Payload[f.arg2StartOffset:f.arg2EndOffset]
+
+	deltaLen := 2 + len(key) + 2 + len(val)
+	if int(f.Header.PayloadSize()) +deltaLen > MaxFramePayloadSize {
+		return fmt.Errorf("frame too large")
+	}
+
+	delta := bufPool.Get().(*bytes.Buffer)
+	writeDataToBuffer(key, delta)
+	writeDataToBuffer(val, delta)
+	f.Payload = append(f.Payload[:f.arg2EndOffset], delta.Bytes()...)
+	f.arg2EndOffset += deltaLen
+
+	arg2PairCount := int(binary.BigEndian.Uint16(arg2Payload[0:2]))
+	binary.BigEndian.PutUint16(arg2Payload[0:2], uint16(arg2PairCount+1))
```

Comment:
So arg2 modification requires the following
 * Update the actual "arg2" which is what this is doing, which is relevant to the Thrift-specific encoding. (Note, we should only ever modify arg2 in this way when format or "as" header == thrift)
 * Update the length of arg2 (which is 2 bytes before arg2 starts)
 * Move arg3
 * Update the size of the frame (as it has likely gotten larger)
 * Update the checksum